OSOE-46: Use WebApplicationFactory directly instead of running the tested app with the dotnet CLI

dministro opened the pull request:
Fixes GH-101
OSOE-46

## Migrating from v3.*

### Preparing WebApplication

1. Prepare the web app to be tested as described in [Basic tests with the default WebApplicationFactory](https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?source=recommendations&view=aspnetcore-6.0#basic-tests-with-the-default-webapplicationfactory).
2. Remove `Lombiq.Tests.UI.AppExtensions` project reference or NuGet package from the web app.
3. Add `Microsoft.Extensions.Configuration.ConfigurationManager` instance directly to `WebApplicationBuilder.Services`.

```diff
using Microsoft.AspNetCore.Builder;
using Microsoft.Extensions.DependencyInjection;
using OrchardCore.Logging;

var builder = WebApplication.CreateBuilder(args);
builder.Host.UseNLogHost();

var configuration = builder.Configuration;

- builder.Services.AddOrchardCms(orchard => orchard.ConfigureUITesting(configuration, enableShortcutsDuringUITesting: true));
+ builder.Services.AddSingleton(configuration);
+ builder.Services.AddOrchardCms();

var app = builder.Build();

app.UseStaticFiles();
app.UseOrchardCore();
app.Run();

+ [SuppressMessage(
+     "Design",
+     "CA1050: Declare types in namespaces",
+     Justification = "As described here(https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-6.0).")]
+ public partial class Program
+ {
+     protected Program()
+     {
+         // Nothing to do here.
+     }
+ }
```

### Preparing UI test project

1. Add a _project reference_ of the web app to be tested to the UI test project.

```diff
  </ItemGroup>

  <ItemGroup>
+    <ProjectReference Include="..\..\src\Lombiq.OSOCE.Web\Lombiq.OSOCE.Web.csproj" />
    <ProjectReference Include="..\..\src\Modules\Lombiq.ChartJs\Lombiq.ChartJs.Tests.UI\Lombiq.ChartJs.Tests.UI.csproj" />
    <ProjectReference Include="..\..\src\Modules\Lombiq.DataTables\Lombiq.DataTables\Tests\Lombiq.DataTables.Tests.UI\Lombiq.DataTables.Tests.UI.csproj" />
    <ProjectReference Include="..\..\src\Modules\Lombiq.HelpfulExtensions\Lombiq.HelpfulExtensions.Tests.UI\Lombiq.HelpfulExtensions.Tests.UI.csproj" />
```

2. Change `OrchardCoreUITestBase` implementation like below. `AppAssemblyPath` is not required anymore.

```diff
namespace Lombiq.OSOCE.Tests.UI;

- public class UITestBase : OrchardCoreUITestBase
+ public class UITestBase : OrchardCoreUITestBase<Program>
{
-     protected override string AppAssemblyPath => WebAppConfigHelper
-         .GetAbsoluteApplicationAssemblyPath("Lombiq.OSOCE.Web", "net6.0");
```

### Breaking changes

There is a breaking change in adding command line arguments to the WebApplication.

    To add command line argument with value, use `InstanceCommandLineArgumentsBuilder.AddWithValue` instead of double call `ArgumentsBuilder.Add`.

    To add command line switch, use `InstanceCommandLineArgumentsBuilder.AddSwitch`.

```diff
                configuration.HtmlValidationConfiguration.RunHtmlValidationAssertionOnAllPageChanges = false;
                configuration.OrchardCoreConfiguration.BeforeAppStart += (_, argsBuilder) =>
                {
-                     argsBuilder.Add("--OrchardCore:OrchardCore_Admin:AdminUrlPrefix").Add("custom-admin");
+                     argsBuilder.AddWithValue("OrchardCore:OrchardCore_Admin:AdminUrlPrefix", "custom-admin");

                    return Task.CompletedTask;
                };
```

### Non breaking changes

#### `Lombiq.Tests.UI.Extensions.ShortcutsUITestContextExtensions`

The following extension methods behave largely the same and their signatures didn't change. Using `WebApplicationFactory` made it possible to use OC services directly instead of using _controller actions_ invoked via browser navigation.

This means that calling the extension methods below don't cause browser navigation any more.

- `SetUserRegistrationTypeAsync`
- `CreateUserAsync`
- `AddUserToRoleAsync`
- `AddPermissionToRoleAsync`
- `EnableFeatureDirectlyAsync`
- `DisableFeatureDirectlyAsync`
- `ExecuteRecipeDirectlyAsync`
- `SelectThemeAsync`
- `GenerateHttpEventUrlAsync`

##### Here is a sample for better understanding:

```diff
    public static async Task EnablePrivacyConsentBannerFeatureAndAcceptPrivacyConsentAsync(this UITestContext context)
    {
        await context.EnablePrivacyConsentBannerFeatureAsync();
-         await context.GoToHomePageAsync();
+         await context.GoToHomePageAsync(onlyIfNotAlreadyThere: false);
        await context.AcceptPrivacyConsentAsync();
        context.Refresh();
    }
```

The original code with the new behavior failed a test, because the browser pointed to the _Home page_ before `await context.EnablePrivacyConsentBannerFeatureAsync()(=> context.EnableFeatureDirectlyAsync({featureId})`. So the `context.EnableFeatureDirectlyAsync` doesn't navigate away, the `await context.GoToHomePageAsync()` call does nothing, and the consent banner doesn't come up.

The solution, in this case, is to call `await context.GoToHomePageAsync(onlyIfNotAlreadyThere: false)`, this result a reload, and the consent banner come up.

Piedone committed:
Basics of WebApplicationFactory usage

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-46

dministro committed:
Adding minimal WebDriverFactory implementation

dministro committed:
Injecting FakeStore

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-46

dministro committed:
Clearing SQLite connection pool

dministro committed:
Fixing ObjectDisposedException in case of restarting Orchard application

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-46

dministro committed:
Fixing CA1063: Provide an overridable implementation of Dispose(bool)

dministro committed:
Fixing the cause of failing tests:
 - BehaviorVueTests.RecipeDataShouldBeDisplayedCorrectly(browser: Chrome)
 - EmailTests.SendingTestEmailShouldWork(browser: Chrome)

dministro committed:
Maybe fixing failing test:
 - Lombiq.Tests.UI.Samples.Tests.ErrorHandlingTests.ErrorDuringSetupShouldHaltTest(browser: Chrome)

dministro committed:
Adding Migration.md

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-46

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-46

BenedekFarkas committed:
Merge branch 'dev' into issue/OSOE-46

dministro committed:
Merge branch 'issue/OSOE-46' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-46

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-46

dministro committed:
Fixing after merge

dministro committed:
Fixing command line arg

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-46

dministro committed:
Renaming AddValue to AddWithValue

dministro committed:
Disabling tests

dministro committed:
Adding logger configuration

dministro committed:
Enabling Lombiq.Tests.UI.Samples tests

dministro committed:
Fixing MonkeyTests

dministro committed:
Fixing MonkeyTests again

dministro committed:
Coding style
Migration guide

dministro committed:
Removing OrchardCoreUITestBase.AppAssemblyPath

dministro committed:
Using OC services directly instead of Shortcut Controllers where possible

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-46

dministro committed:
Using OC services directly instead of Shortcut Controllers where possible, the rest

dministro committed:
Removing unused variable

dministro committed:
Injecting fake HttpContext before calling OC services directly

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-46

dministro committed:
Using OC services directly instead of Shortcut Controllers where possible, one more

dministro committed:
Fixing docs

dministro committed:
Using Lombiq.Tests.UI.Samples test classes from Lombiq.OSOCE.Tests.UI as a consumer.

dministro committed:
Removing test sdk from Lombiq.Tests.UI.Samples.csproj

dministro committed:
Adding Lombiq.Tests as a NuGet package

dministro committed:
Revert "Removing test sdk from Lombiq.Tests.UI.Samples.csproj"

dministro committed:
Moving OrchardCoreBuilderExtensions to Lombiq.Tests.UI
Updating docs

dministro committed:
Fixing docs

dministro committed:
Fixing docs

dministro committed:
Removing possible dead-lock

dministro committed:
Fixing razor runtime compilation related issue

dministro committed:
Just for a test

dministro committed:
Fixing possible dead lock

dministro committed:
Removing unnecessary workaround for razor runtime compilation

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-46

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-46

dministro committed:
Updating docs

dministro committed:
Possible fix of real deadlock
Updating docs

dministro committed:
Adding IServiceProviderInstance to fake HttpContext

dministro committed:
Trying to avoid adding razor runtime view compilation

dministro committed:
Updating docs

Piedone commented:
Great migration docs, BTW!

sarahelsaig commented on Lombiq.Tests.UI/Docs/Migration.md:
```diff
@@ -0,0 +1,120 @@
+### Preparing WebApplication
+
+1. Prepare the web app uneder test as described in [Basic tests with the default WebApplicationFactory](https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?source=recommendations&view=aspnetcore-6.0#basic-tests-with-the-default-webapplicationfactory).
+2. Remove `Lombiq.Tests.UI.AppExtensions` project reference or NuGet package from the web app under test.
```
```suggestion
1. Prepare the web app to be tested as described in [Basic tests with the default WebApplicationFactory](https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?source=recommendations&view=aspnetcore-6.0#basic-tests-with-the-default-webapplicationfactory).
2. Remove `Lombiq.Tests.UI.AppExtensions` project reference or NuGet package from the web app.
```

sarahelsaig commented on Lombiq.Tests.UI/Docs/Migration.md:
```diff
@@ -0,0 +1,120 @@
+app.UseOrchardCore();
+app.Run();
+
++ // As described here(https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-6.0).
```
```suggestion
+ // As described here (https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-6.0).
```

sarahelsaig commented on Lombiq.Tests.UI/Docs/Migration.md:
```diff
@@ -0,0 +1,120 @@
+
+### Preparing UI test project
+
+1. Add the web app project under test as a `Project reference` to UI test project.
```
Only use backticks for code.

```suggestion
1. Add a _project reference_ of the web app to be tested to the UI test project.
```

sarahelsaig commented on Lombiq.Tests.UI/Docs/Migration.md:
```diff
@@ -0,0 +1,120 @@
+
+### Breaking changes
+
+There is a breaking change in adding command line arguments to the WebApplication.
```
It's not so great that people have to have read this document to know about this. Please add a stub `Add(string)` method to `InstanceCommandLineArgumentsBuilder` that's marked `[Obsolete]` and throws `NotSupportedException when called`.

sarahelsaig commented on Lombiq.Tests.UI/Docs/Migration.md:
```diff
@@ -0,0 +1,120 @@
+
+#### `Lombiq.Tests.UI.Extensions.ShortcutsUITestContextExtensions`
+
+The following extension methods result in the same and no breaking change in the signatures, but they behave a bit differently. Using `WebApplicationFactory` opened the possibility to use OC services directly instead of using `Controller Actions` invoked by browser navigation.
```
```suggestion
The following extension methods behave largely the same and their signatures didn't change. Using `WebApplicationFactory` made it possible to use OC services directly instead of using _controller actions_ invoked via browser navigation.
```

sarahelsaig commented on Lombiq.Tests.UI/Docs/Migration.md:
```diff
@@ -0,0 +1,120 @@
+
+The following extension methods result in the same and no breaking change in the signatures, but they behave a bit differently. Using `WebApplicationFactory` opened the possibility to use OC services directly instead of using `Controller Actions` invoked by browser navigation.
+
+This means that calling the extension methods below no longer results in browser navigation.
```
```suggestion
This means that calling the extension methods below don't cause browser navigation any more.
```

sarahelsaig commented on Lombiq.Tests.UI/Docs/Migration.md:
```diff
@@ -0,0 +1,120 @@
+- `SelectThemeAsync`
+- `GenerateHttpEventUrlAsync`
+
+_Here is a sample for better understanding_:
```
Why is this italicized?

sarahelsaig commented on Lombiq.Tests.UI/Docs/Migration.md:
````diff
@@ -0,0 +1,120 @@
+    }
+```
+
+The original code with the new behavior resulted in a failing test, because the browser pointed to the `Home page` before `await context.EnablePrivacyConsentBannerFeatureAsync()(=> context.EnableFeatureDirectlyAsync({featureId})`, the `context.EnableFeatureDirectlyAsync` not navigates away, so finaly the `await context.GoToHomePageAsync()` call do nothing and the consent banner din't come up.
````
```suggestion
The original code with the new behavior failed a test, because the browser pointed to the _Home page_ before `await context.EnablePrivacyConsentBannerFeatureAsync()(=> context.EnableFeatureDirectlyAsync({featureId})`. So the `context.EnableFeatureDirectlyAsync` doesn't navigate away, the `await context.GoToHomePageAsync()` call does nothing, and the consent banner doesn't come up.
```

sarahelsaig commented on Lombiq.Tests.UI/Docs/TestableOrchardCoreApps.md:
```diff
@@ -6,26 +6,19 @@ Tips on making specific features testable are under the ["Creating tests" page](
 - In your web project do the following:
-  1. Add a reference to `Lombiq.Tests.UI.AppExtensions` (either from NuGet or as a Git submodule).
-  2. Allow configuration of the app when launched for testing with the following piece of code in the app's `Startup` class:
+  1. Prepare the web app uneder test as described in [Basic tests with the default WebApplicationFactory](https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?source=recommendations&view=aspnetcore-6.0#basic-tests-with-the-default-webapplicationfactory).
```
```suggestion
  1. Prepare the web app to be testes as described in [Basic tests with the default WebApplicationFactory](https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?source=recommendations&view=aspnetcore-6.0#basic-tests-with-the-default-webapplicationfactory).
```

sarahelsaig commented on Lombiq.Tests.UI/Extensions/VisualVerificationUITestContextExtensions.cs:
```diff
@@ -637,8 +639,46 @@ public static class VisualVerificationUITestContextExtensions
+    private static string GetMethodeName(this EnhancedStackFrame frame)
+    {
+        var methodName = frame.MethodInfo.Name;
+        var inheriteedMethod = new Regex("^<(?<method>.*)>.*$", RegexOptions.ExplicitCapture);
+        if (inheriteedMethod.IsMatch(methodName))
+        {
+            methodName = inheriteedMethod.Match(methodName).Groups["method"].Value;
+        }
```
What is this?

dministro replied:
Commented in code

sarahelsaig commented on Lombiq.Tests.UI/Models/InstanceCommandLineArgs.cs:
```diff
@@ -0,0 +1,34 @@
+
+    public InstanceCommandLineArgumentsBuilder AddWithValue<T>(string key, T value)
+    {
+        _arguments.Add($"{PrepareArg(key)}={value}");
```
Wrap the string interpolation with `FormattableString.Invariant()` because value may be an `IFormattable` such as `int`.

sarahelsaig commented on Lombiq.Tests.UI/Models/InstanceCommandLineArgs.cs:
```diff
@@ -0,0 +1,34 @@
+
+    private static string PrepareArg(string argument)
+    {
+        if (!argument.StartsWith("--", System.StringComparison.InvariantCultureIgnoreCase))
```
This handles `--argument`, but not `-argument` which is also valid. So just doing `.TrimStart('-')` would be better.

sarahelsaig commented on Lombiq.Tests.UI/Services/OrchardCoreHosting/OrchardApplicationFactory.cs:
```diff
@@ -0,0 +1,122 @@
+                .LastOrDefault(descriptor => descriptor.ServiceType == typeof(OrchardCoreBuilder))?
+                .ImplementationInstance as OrchardCoreBuilder
+                ?? throw new InvalidOperationException(
+                    "Pease call WebApplicationBuilder.Services.AddOrchardCms() in your Program.cs!");
+        var configuration = services
+                .LastOrDefault(descriptor => descriptor.ServiceType == typeof(ConfigurationManager))?
+                .ImplementationInstance as ConfigurationManager
+                ?? throw new InvalidOperationException(
+                    "Pease add ConfigurationManager instance to WebApplicationBuilder.Services in your Program.cs!");
```
Use `nameof`.

dministro committed:
Update Lombiq.Tests.UI/Docs/Migration.md

Co-authored-by: Dávid El-Saig <david.el-saig@lombiq.com>

dministro committed:
Update Lombiq.Tests.UI/Docs/Migration.md

Co-authored-by: Dávid El-Saig <david.el-saig@lombiq.com>

dministro committed:
Update Lombiq.Tests.UI/Docs/Migration.md

Co-authored-by: Dávid El-Saig <david.el-saig@lombiq.com>

dministro committed:
Update Lombiq.Tests.UI/Docs/Migration.md

Co-authored-by: Dávid El-Saig <david.el-saig@lombiq.com>

dministro committed:
Update Lombiq.Tests.UI/Docs/Migration.md

Co-authored-by: Dávid El-Saig <david.el-saig@lombiq.com>

dministro committed:
Update Lombiq.Tests.UI/Docs/Migration.md

Co-authored-by: Dávid El-Saig <david.el-saig@lombiq.com>

dministro committed:
Update Lombiq.Tests.UI/Docs/TestableOrchardCoreApps.md

Co-authored-by: Dávid El-Saig <david.el-saig@lombiq.com>

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-46

dministro committed:
Updating docs
Adding comments
Adding InstanceCommandLineArgumentsBuilder.Add with Obsolate attribute

dministro committed:
Updating Lombiq.Tests Nuget package version

sarahelsaig commented on Lombiq.Tests.UI/Models/InstanceCommandLineArgs.cs:
```diff
@@ -0,0 +1,31 @@
+    [Obsolete("Use AddSwitch or AddWithValue instead.")]
+    public InstanceCommandLineArgumentsBuilder Add(string value) => throw new NotSupportedException();
+
+    private static string PrepareArg(string argument)
+        => $"--{argument.TrimStart('-')}";
```
```suggestion
    private static string PrepareArg(string argument) =>
        $"--{argument.TrimStart('-')}";
```

sarahelsaig committed:
Update Lombiq.Tests.UI/Models/InstanceCommandLineArgs.cs

sarahelsaig commented on Lombiq.Tests.UI/Services/OrchardCoreHosting/OrchardApplicationFactory.cs:
```diff
@@ -0,0 +1,122 @@
+
+    private void ConfigureTestServices(IServiceCollection services)
+    {
+        var builder = services
+                .LastOrDefault(descriptor => descriptor.ServiceType == typeof(OrchardCoreBuilder))?
+                .ImplementationInstance as OrchardCoreBuilder
+                ?? throw new InvalidOperationException(
+                    "Pease call WebApplicationBuilder.Services.AddOrchardCms() in your Program.cs!");
+        var configuration = services
+                .LastOrDefault(descriptor => descriptor.ServiceType == typeof(ConfigurationManager))?
+                .ImplementationInstance as ConfigurationManager
+                ?? throw new InvalidOperationException(
+                    $"Pease add {nameof(ConfigurationManager)} instance to WebApplicationBuilder.Services in your Program.cs!");
```
```suggestion
        var builder = services
                .LastOrDefault(descriptor => descriptor.ServiceType == typeof(OrchardCoreBuilder))?
                .ImplementationInstance as OrchardCoreBuilder
                ?? throw new InvalidOperationException(
                    "Please call WebApplicationBuilder.Services.AddOrchardCms() in your Program.cs!");
        var configuration = services
                .LastOrDefault(descriptor => descriptor.ServiceType == typeof(ConfigurationManager))?
                .ImplementationInstance as ConfigurationManager
                ?? throw new InvalidOperationException(
                    $"Please add {nameof(ConfigurationManager)} instance to WebApplicationBuilder.Services in your Program.cs!");
```

sarahelsaig committed:
Update Lombiq.Tests.UI/Services/OrchardCoreHosting/OrchardApplicationFactory.cs
